src/github.com/getlantern/detour/detect.go

```diff
@@ -0,0 +1,65 @@
+package detour
+
+import (
+	"bytes"
+	"net"
+	"syscall"
+)
+
+type Detector struct {
+	CheckConn    func(net.Conn) bool
+	CheckError   func(error) bool
+	CheckContent func([]byte) bool
+}
+
+var detectors = make(map[string]*Detector)
+
+var iranRedirectAddr = "10.10.34.34:80"
+
+func init() {
+	http403 := []byte("HTTP/1.1 403 Forbidden")
+	iranIFrame := []byte(`<iframe src="http://10.10.34.34`)
+	// see tests and https://github.com/getlantern/lantern/issues/2099#issuecomment-78015418
+	// for the facts behind detection rules for Iran
+	detectors["IR"] = &Detector{
+		CheckConn: func(c net.Conn) bool {
+			if ra := c.RemoteAddr(); ra != nil {
+				return ra.String() == iranRedirectAddr
+			}
+			return false
+		},
+		CheckContent: func(b []byte) bool {
+			return bytes.HasPrefix(b, http403) && bytes.Contains(b, iranIFrame)
```

Comment:
Also, I think I heard someone in Valencia mention that Iran has started redirecting to sites other than the standard redirect (even sometimes real sites).  Detecting that may be hard, but that's probably okay, since even if detour only detects blocking sometimes, that's better than never.